### PR TITLE
lib/assets: Allow assets to remain uncompressed

### PR DESCRIPTION
**Pull request by greatroar**
All compiled-in assets are currently compressed, but six of them become larger because of this, so the compression/decompression step is a waste of time and space: two WOFF fonts, three PNG images and a tiny JavaScript file.

This PR adds support for uncompressed assets. The generator script decides on the format depending on whether compression makes it bigger or smaller. Also, the compression level is set to BestCompression. The binary gets 4kB smaller on Linux/amd64 as a result.

### CODE REVIEW

File: lib/assets/assets.go [hunk 1]
```diff
@@ -53,19 +53,28 @@ func Serve(w http.ResponseWriter, r *http.Request, asset Asset) {
...
 		var gr *gzip.Reader
-		gr, _ = gzip.NewReader(strings.NewReader(asset.ContentGz))
+		gr, _ = gzip.NewReader(strings.NewReader(asset.Content))
 		io.Copy(w, gr)
 		gr.Close()
 	}
 }
 
+func gzipped(content string) bool {
+	const gzmagic = "\x1f\x8b" // RFC 1952, §2.3.1.
```
Comment (AudriusButkevicius) [inline lib/assets/assets.go:74]: We should just have a flag/enum for the compression used, if any.

Comment (calmh) [reply lib/assets/assets.go:74]: At which point we can also add an `length int` and serve that as `Content-Length` to clients not speaking gzip.

**Comment by calmh** [comment]
> The binary gets 4kB smaller on Linux/amd64 as a result.

That's almost 0.02% percent of an improvement!

**Comment by greatroar** [comment]
Yeah, this is a nano-optimization in terms of space. Skipping the decompression in the browser and giving it a proper Content-Length are more important than the space saving, but those effects are much harder to measure.

Btw., I've toyed around with [go-zopfli](https://github.com/foobaz/go-zopfli) and it can shave off a few percent.

File: lib/assets/assets.go (hunk 1 above)
Comment (greatroar) [reply lib/assets/assets.go:74]: That would be nicer, but the code for it is much more complicated. The generated code and the client in lib/assets need some common type to communicate bool+string.

Comment (calmh) [reply lib/assets/assets.go:74]: The `Asset` struct?

Comment (greatroar) [reply lib/assets/assets.go]: Yes. Turned out to be easier than I thought, see updated code.

**Comment by greatroar** [comment]
Looks like the failing tests are due to go generate not being run. They work fine on my machine.

File: lib/assets/assets.go [hunk 2]
```diff
@@ -53,14 +55,18 @@ func Serve(w http.ResponseWriter, r *http.Request, asset Asset) {
...
+
+	switch {
+	case !asset.Gzipped:
+		io.WriteString(w, asset.Content)
+	case strings.Contains(r.Header.Get("Accept-Encoding"), "gzip"):
 		header.Set("Content-Encoding", "gzip")
-		header.Set("Content-Length", strconv.Itoa(len(asset.ContentGz)))
-		io.WriteString(w, asset.ContentGz)
-	} else {
+		io.WriteString(w, asset.Content)
```
Comment (AudriusButkevicius) [inline lib/assets/assets.go:65]: Content-length gone?

Comment (calmh) [reply lib/assets/assets.go:65]: It's above

**Comment by calmh** [comment]
Test failures seem genuine though

```
# github.com/syncthing/syncthing/lib/api [github.com/syncthing/syncthing/lib/api.test]
lib/api/api_test.go:167:3: cannot use map[string]string literal (type map[string]string) as type map[string]assets.Asset in field value
FAIL	github.com/syncthing/syncthing/lib/api [build failed]

--- FAIL: TestServeGzip (0.00s)
    assets_test.go:86: unexpected content "\x1f\x8b\b\x00\x00\x00\x00\x00\x00\xff\xb2\xc9(\xc9ͱ\xf3H\xcd\xc9\xc9\xd7Q(\xcf/\xcaIQ\xb4\xd1\a\x8b\x01\x02\x00\x00\xff\xff/\xa3B'\x1a\x00\x00\x00"
FAIL
```

File: lib/assets/assets.go (hunk 2 above)
Comment (calmh) [reply lib/assets/assets.go:65]: Wait. We're now sending the length of the uncompressed asset with the compressed data? does that work?

Comment (greatroar) [reply lib/assets/assets.go:65]: No. Sorry, I was mistaken. It turns out [Content-Length is the encoded length](https://stackoverflow.com/a/3819303).

File: lib/assets/assets.go [hunk 3]
```diff
@@ -53,14 +54,18 @@ func Serve(w http.ResponseWriter, r *http.Request, asset Asset) {
...
+	default:
 		// gunzip for browsers that don't want gzip.
 		var gr *gzip.Reader
-		gr, _ = gzip.NewReader(strings.NewReader(asset.ContentGz))
+		gr, _ = gzip.NewReader(strings.NewReader(asset.Content))
```
Comment (calmh) [inline lib/assets/assets.go:68]: wrong content-length here?

Comment (greatroar) [reply lib/assets/assets.go:70]: Yes. Should be fixed now, and I added an assertion to the test.

**Pull request merged**